### PR TITLE
Add engine-v2 graph

### PR DESCRIPTION
**Pull request by Finistere**
Adding a base structure to work on, will definitely iterate a lot on it.
It's almost a copy paste from the federated-graph for now.

I don't think unit tests for the conversion itself are meaningful. This should
be tested through integration tests like composition. But instead of
checking a SDL output we would check introspection and that the engine
behaves as expected.

# Description

Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.

# Type of change

- [ ] 💔 Breaking
- [ ] 🚀 Feature
- [ ] 🐛 Fix
- [ ] 🛠️ Tooling
- [ ] 🔨 Refactoring
- [ ] 🧪 Test
- [ ] 📦 Dependency
- [ ] 📖 Requires documentation update

### CODE REVIEW

File: engine/crates/engine-v2/src/graph/conversion.rs [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+                .iter()
+                .enumerate()
+                .find_map(|(id, object)| {
+                    if graph[object.name] == name {
+                        Some(ObjectId(id))
+                    } else {
+                        None
+                    }
+                })
+                .expect(&format!("{name} object not found"))
```
[the remark this answers is not in the document]
Comment (Finistere) [reply engine/crates/engine-v2/src/graph/conversion.rs:19]: it's only called twice hence my comment, but fair point.^^

File: engine/crates/engine-v2/src/graph/conversion.rs [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        };
+        let query_object_id = find_object("Query");
+        let mutation_object_id = find_object("Mutation");
```
[the remark this answers is not in the document]
Comment (Finistere) [reply engine/crates/engine-v2/src/graph/conversion.rs:22]: Ah was convinced the spec specified those names, but you're right

File: engine/crates/engine-v2/src/graph/mod.rs [hunk 3]
```diff
@@ -0,0 +1,288 @@
+mod conversion;
+
+/// This does NOT need to be backwards compatible. We'll probably cache it for performance, but it is not
+/// the source of truth. If the cache is stale we would just re-create this Graph from its source:
+/// federated_graph::FederatedGraph.
+#[allow(dead_code)]
+pub struct Graph {
```
[the remark this answers is not in the document]
Comment (Finistere) [reply engine/crates/engine-v2/src/graph/mod.rs:7]: `RuntimeGraph` / `EngineGraph` maybe?

File: engine/crates/engine-v2/src/parser.rs [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+use crate::graph::{FieldId, FieldTypeId, Graph, InterfaceId, ObjectId, StringId};
+
+#[allow(dead_code)]
+pub struct VariableId(usize);
+
+#[allow(dead_code)]
+pub enum TypeCondition {
+    Interface(InterfaceId),
+    Object(ObjectId),
+}
```
[the remark this answers is not in the document]
Comment (Finistere) [reply engine/crates/engine-v2/src/parser.rs:14]: good catch

**Review by obmarg — APPROVED** [review]

**Pull request merged**